3rd_homework

mypetproject opened the pull request:
Добавлены два фрагмента в MainActivity. В NotesListFragment имплементирован RecyclerView. Результат клика по элементам RecyclerView отображается в NoteDataFragment. NotesModel теперь хранит MutableList с записями.

Commit:
Добавлены два фрагмента в MainActivity. В NotesListFragment имплементирован RecyclerView. Результат клика по элементам RecyclerView отображается в NoteDataFragment. NotesModel теперь хранит MutableList с записями.

BSCAndroid commented on app/src/main/java/com/example/bschomework/NoteData.kt:
```diff
@@ -0,0 +1,6 @@
+import java.io.Serializable
+import java.util.*
+
+data class NoteData(var header: String, var note: String, var date: Date) : Serializable
```
Лучше Parcelable вместо Serializable

BSCAndroid commented on app/src/main/java/com/example/bschomework/NotesListAdapter.kt:
```diff
@@ -0,0 +1,43 @@
+    override fun getItemCount() = notes.size
+
+    inner class NoteViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        val header: TextView = view.findViewById(R.id.header_rv)
```
Можно обойтись без лишней переменной

BSCAndroid commented on app/src/main/java/com/example/bschomework/fragments/NoteDataFragment.kt:
```diff
@@ -0,0 +1,36 @@
+        setFragmentResultListener("dataForNoteDataFragment") { _, bundle ->
+            val result = bundle.getSerializable("noteData") as NoteData
+
+            binding.header.text = result.header
```
Можно использовать скоуп-функцию

BSCAndroid commented on app/src/main/java/com/example/bschomework/fragments/NoteDataFragment.kt:
```diff
@@ -0,0 +1,36 @@
+
+        val binding = FragmentNoteDataBinding.inflate(inflater, container, false)
+
+        setFragmentResultListener("dataForNoteDataFragment") { _, bundle ->
```
Все вот такие строки должны находиться в companion object во фрагменте, активити.

BSCAndroid commented on app/src/main/java/com/example/bschomework/fragments/NoteDataFragment.kt:
```diff
@@ -0,0 +1,36 @@
+        val binding = FragmentNoteDataBinding.inflate(inflater, container, false)
+
+        setFragmentResultListener("dataForNoteDataFragment") { _, bundle ->
+            val result = bundle.getSerializable("noteData") as NoteData
```
Все вот такие строки должны находиться в константах в companion object во фрагменте, активити

BSCAndroid commented on app/src/main/java/com/example/bschomework/NotesListAdapter.kt:
```diff
@@ -0,0 +1,43 @@
+    private val onItemClick: ((NoteData, Int) -> Unit)?
+) : RecyclerView.Adapter<NotesListAdapter.NoteViewHolder>() {
+
+    var selectedPosition = RecyclerView.NO_POSITION
```
Не стоит менять переменную извне напрямую

BSCAndroid commented on app/src/main/java/com/example/bschomework/NotesListAdapter.kt:
```diff
@@ -0,0 +1,43 @@
+
+        init {
+            view.setOnClickListener {
+                onItemClick?.invoke(notes[absoluteAdapterPosition], absoluteAdapterPosition)
```
onItemClick лучше передать в bind и использовать его там

Commit:
Serializable заменён на Parcelable. Data Binding добавлен в notes_list_item.xml. В NoteDataFragment использованы скоуп-функции.
Ключи в setFragmentResultListener и setFragmentResult вынесены в companion object в NoteDataFragment. Переменная selectedPosition теперь privat и изменяется только внутри NotesListAdapter.

BSCAndroid commented on app/src/main/java/com/example/bschomework/NotesListAdapter.kt:
```diff
@@ -27,17 +26,21 @@ class NotesListAdapter(
-            }
+        fun bind(note: NoteData) {
+            binding.headerRv.text = note.header
+            itemView.setOnClickListener(this)
```
Зачем так сложно? :) Можно просто onItemClick передать в bind и все. Это делается для того, чтобы холдер был независим от адаптера в целом и не использовал его внутренние члены, а получал нужное через параметр.

Commit:
В NoteViewHolder onClick перенесён в bind.

Commit:
В MainActivity.kt отображается список заметок. При нажатии на заметку открывается фрагмент с деталями заметки в новом активити.

Commit:
NoteDataFragment.kt перенесён в MainActivity.kt

Commit:
Фрагмент добавляется в back stack при замене в контейнере.
